### ISSUE TITLE
Recording timing issues

### ISSUE THREAD

**Issue by O5ten** [aerokube/selenoid#313]
Hey! 

First off, i love Selenoid, it's just plain awesome to have an out-of-the-box docker-based selenium-grid for the rest of us. It saved us a bunch of time and headache integrating selenium system tests into an ancient Jenkins setup. So, thank you! 

**To the point**
We have the new video-feature turned on and have a wget towards the video url http://grid:444/video/${JOB_NAME}-${BUILD_NUMBER}.mp4 followed by a DELETE to store the video with the CI jobs instead.

**Issue** 
When a very short job, say an early fail, is triggered then the recorder usually turns up with zero-to-almost nothing in size and the recording is impossible to view so i have to put a sleep on it. I assume this is because the recording simply hasn't been completed yet or that the first buffer hasn't been emptied into the file yet. We've also noticed that the recording is viewable in the browser on the grid, but technically never on the Jenkins-machine. My guess is it is lacking some video encoding EOF that the browser can't handle, i have no idea though. 

**But regarding timing**
Is it possible perhaps to make sure that the selenium-session isn't closed until the recording is completed, so we'd avoid the timeout alltogether and that we can trust that the recording on the grid is completed when the session is?

**Comment by vania-pooh** [comment]
@O5ten currently it works like the following.
1) When you specify `enableVideo: true` in your tests - an additional container with FFMpeg is started and attached to running browser container using Docker links. This is needed to capture the video via X-server protocol using `ffmpeg -f x11grab <...>` feature.
2) When you request to stop browser session we proxy your request to container with browser and return the answer.
3) The rest of cleanup actions are currently being done asynchronously in parallel thread (goroutine). This cleanup includes removing browser container as well as sending SIGTERM to video container. Only when receiving such signal `ffmpeg` exits correctly and flushes all remaining video data to file. If you do not specify `videoName` Selenoid additionally renames temporary file by its session ID.

So yes, in current implementation a situation may occur when your tests have already finished but video is being saved. We need to think whether making all this cleanup synchronous can slow down things.

@aandryashin any ideas about it?

**Comment by Maratori** [comment]
The solution for me was not to use custom videoName. If so, selenoid writes file "selenoid<session-id>.mp4" and after finish recording renames it to "<session-id>.mp4". So I can wait while such file appears (through direct access or http).

**Comment by Maratori** [comment]
I see two solutions to improve the situation:
1. Write file "selenoid&lt;session-id&gt;.mp4" anyway and then rename it either to "&lt;session-id&gt;.mp4" or to requested name.
2. Add new capability. Something like waitVideoFinish=true. In that case make process sycronous. Or just send signal that browser is closed to the test only after video is finished.

**Comment by O5ten** [comment]
i'm not a fan of having to figure out the session-id at the Jenkins-level (where i download the files from). 

The problem right now is that there's no way to _know_ that the recording is completed. 
1. A prefix as @Maratori said, a tempfile-prefix-<session-id-or-videoName>.mp4 then i could make the script wait for that to happen if its asychronous. 
2. A synchronous approach where the selenoid session isn't stopped until the recording is finished. 
3. The file being stored somewhere else or hidden from access with HTTP until the video-recorder is finished. (This would also be possible with a capability i assume hideVideoUntilFinished=true or similar)

I'm fine with either approach as long as i can have a safe way to download the video without static delays. :)

**Comment by vania-pooh** [comment]
@O5ten we could change the logic so `videoName` capability works in the same way like without it. That is to say `http://selenoid-host.example.com:4444/video/<video-name>.mp4` will return 200 and video contents only when video is fully written. So we'll use temporary file name always and rename to either `<session-id>.mp4` or `<video-name>.mp4` depending on capabilities. In that case waiting for video could be something like (pseudocode):
```
while true {
    curl -I http://selenoid-host.example.com:4444/video/<video-name>.mp4
    if not 200 then
        sleep 500
    else
        break
    endif
}
```

**Comment by vania-pooh** [comment]
I can state that video feature is a working prototype. So any improvement proposals like this are welcome.

**Comment by O5ten** [comment]
@vania-pooh That's an ever better solution, tbh. 

And we're happy to be test-driving the prototype. 
It's already a great help for us when debugging the nightly tests.

**Comment by vania-pooh** [comment]
Ok, will be implemented soon.

**Comment by vania-pooh** [comment]
Fixed. Currently available in `aerokube/selenoid:latest`. Will be released in version `1.4.2`.
